lib/AST/ASTImporter.cpp

```diff
@@ -2570,7 +2570,10 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
     // "void f(); void f() { f(); }"
     // This is because in the definition the CallExpression refers to the
     // definition, not to the prototype.
-    Importer.Imported(const_cast<FunctionDecl*>(FromDefinition), ToFunction);
+    if (!Importer.GetAlreadyImportedOrNull(
+            const_cast<FunctionDecl *>(FromDefinition))) {
```

Comment:
4 spaces of indentation is sufficient here? (2 less than now)
And { } could be left out from this `if`.
Otherwise it is the same correction that I tried already, seems to work OK.

Reply:
Maybe we could hoist the const cast to not repeat it?

Also, I am wondering what do we want to fix here? If the FromDefinition is already imported, do we need to set the body bellow or could we just early return?

If we early return, we already created a function without a body in the to context. If we do not early return do we have two functions with a body? This sounds troublesome. Correct me if I'm wrong.

Reply:
If the function body is already imported we should find it at the search for "similar function" at start of the import and return. I think it may happen that the body definition is imported after that search like the "Check recursive import." case (in this case, not the `D` itself, but another definition is imported that contains the body). So at the place of "Check recursive import" it could be checked too if the `FromDefinition` is imported already. If yes we can return because the function body is already there. But it might be wrong to call `Imported(D, ToDefinition)` because the `Imported` copies some attribute values from the `D` to the `ToDefinition` that may be not good in this case (these values should be copied from the `FromDefinition` not from `D`).

Reply:
The body might have been imported after the search when we import the describing template or when we import the params. My concern is either having multiple declarations with a body or having function declaration without the body imported. If we just skip the `imported` call we might end up having the former, if we early return we might end up having the latter.

Reply:
Actually, the latter is not be possible. Since that would mean we could not find the definition with lookup, but the definition is already imported, this is a contradiction.
Thus, my suggestion is:
```
    if (!Importer.GetAlreadyImportedOrNull(
            const_cast<FunctionDecl *>(FromDefinition))) {
      Importer.Imported(const_cast<FunctionDecl *>(FromDefinition), ToFunction);
    } else {
      llvm_unreachable("Lookup could not find already imported definition!");
}

```

Reply:
There is no guarantee we do not import the function definition *after* the lookup did not find anything useufl. In fact, we do for example in cases of templated function definitions.

Reply:
If unreachable fires that means we have a flaw in the lookup logic.

Reply:
Let's add this assert and see how often it fires on xerces and check if those are really lookup related problems.

Reply:
Ok, I am going to do that measurement and will come back with the results.

Reply:
The recursive function case is solved by the added `Imported` call.
I think that it could be better to check at the start of `VisitFunctionDecl` if the `D` has a body defined somewhere else and if yes import that function instead (and call `Imported` with correct values).

Reply:
I had an early attempt to do exactly that a few weeks ago, but that caused an infinite recursion.
Still we changed the a lot since then, so it is worth to try that cleaner attempt again. I'll check that.

Reply:
Now I think that it is bad to do the check at start, it can cause infinite recursion (if during the import of the found definition the same FunctionDecl is encountered again).

But this thing can be done:
```
    if (!Importer.GetAlreadyImportedOrNull(
            const_cast<FunctionDecl *>(FromDefinition)) || FromDefinition == D) {
      Importer.Imported(const_cast<FunctionDecl *>(FromDefinition), ToFunction);
      if (Stmt *ToBody = Importer.Import(FromBody)) {
        ToFunction->setBody(ToBody);
      }
```
The body is imported only if there was no already existing definition (otherwise the body should exist in the other definition).

Reply:
With @Xazax-hun we had a discussion that maybe squashing the definition and the prototype is not the proper way of importing redecl chains. See issue #327. Perhaps we should import the whole chain rather squashing.

Reply:
A solution like this does not cause assert on recursive functions and imports the body only at the original place:
```
  if (Stmt *FromBody = D->getBody(FromDefinition)) {
    if (FromDefinition == D) {
      if (Stmt *ToBody = Importer.Import(FromBody)) {
        ToFunction->setBody(ToBody);
      }
    }
  }
```

Reply:
That is true, but then importing the prototype would fail to import the definition. I am working on another approach now, which will import the whole redecl chain whenever any member of the chain is imported.

Reply:
I tried the code with `llvm_unreachable` and it fires in 334/348. So yes, looks like we have a lookup related problem. However, if we hadn't import something twice than the wrong lookup code would have never been fired. Currently I am working on the second part of the problem.